### PR TITLE
Added Go deployment script

### PR DESCRIPTION
**Pull request by shrimpy**
Deployment script:
1) Create go folder structure in %DEPLOYMENT_TEMP%
2) Kudysync source to %DEPLOYMENT_TEMP%\src
3) Kudysync files other than *.go from %DEPLOYMENT_TEMP%\src to %DEPLOYMENT_TARGET%
4) Resolve build dependency for Go app
5) Build Go app and place output exe to %DEPLOYMENT_TARGET%

Sample continues deployment log (testing in local machine):
````bash
Counting objects: 4, done.
Delta compression using up to 12 threads.
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 288 bytes | 0 bytes/s, done.
Total 3 (delta 1), reused 0 (delta 0)
remote: Updating branch 'master'.
remote: Updating submodules.
remote: Preparing deployment for commit id '16bde249e4'.
remote: Generating deployment script.
remote: Running deployment command...
remote: Handling Go deployment.
remote: Prepare workspace
remote: GOROOT F:\Go
remote: Creating GOPATH\bin C:\Users\go3\AppData\Local\Temp\b532be25-b118-4992-89af-40ed1456b2d3\gopath\bin
remote: Creating GOPATH\pkg C:\Users\go3\AppData\Local\Temp\b532be25-b118-4992-89af-40ed1456b2d3\gopath\pkg
remote: Creating GOPATH\src C:\Users\go3\AppData\Local\Temp\b532be25-b118-4992-89af-40ed1456b2d3\gopath\src
remote: Creating C:\Users\go3\AppData\Local\Temp\b532be25-b118-4992-89af-40ed1456b2d3\gopath\src\azureapp
remote: Copy source code to Go workspace
remote:
remote: -------------------------------------------------------------------------------
remote:    ROBOCOPY     ::     Robust File Copy for Windows
remote: -------------------------------------------------------------------------------
remote:
remote:   Started : Wednesday, April 22, 2015 10:31:43 AM
remote:    Source : E:\kudu\apps\go3\site\repository\
remote:      Dest : C:\Users\go3\AppData\Local\Temp\b532be25-b118-4992-89af-40ed1456b2d3\gopath\src\azureapp\
remote:
remote:     Files : *.*
remote:
remote: Exc Files : .deployment
remote:             deploy.cmd
remote:
remote:  Exc Dirs : .git
remote:             .hg
remote:
remote:   Options : *.* /NDL /NFL /S /E /DCOPY:DA /COPY:DAT /NP /R:1000000 /W:30
remote:
remote: ------------------------------------------------------------------------------
remote:
remote:
remote: ------------------------------------------------------------------------------
remote:
remote:                Total    Copied   Skipped  Mismatch    FAILED    Extras
remote:     Dirs :         3         1         1         0         0         0
remote:    Files :        13        11         2         0         0         0
remote:    Bytes :    22.4 k    22.4 k         4         0         0         0
remote:    Times :   0:00:00   0:00:00                       0:00:00   0:00:00
remote:
remote:
remote:    Speed :             2297900 Bytes/sec.
remote:    Speed :             131.486 MegaBytes/min.
remote:    Ended : Wednesday, April 22, 2015 10:31:43 AM
remote:
remote: Resolving dependencies
remote: ....
remote: Building Go app to produce exe file
remote: Copy static files to deployment
remote: KuduSync.NET from: 'E:\kudu\apps\go3\site\repository' to: 'E:\kudu\apps\go3\site\wwwroot'
remote: Deleting file: 'index.txt'
remote: Copying file: 'index1.txt'
remote: Copying file: 'index2.txt'
remote: Finished successfully.
remote: Deployment successful.
To http://localhost:7952/go3.git
   35bbc91..16bde24  master -> master
````

### CODE REVIEW

**Comment by amitapl** [comment]
For step 2 why not use simple copy/xcopy?

**Comment by shrimpy** [comment]
when user delete some files in later deployment, we would like to remove those files too.  that is why i update to use kudusync.

**Comment by amitapl** [comment]
as this is a temporary directory it will get cleaned up.
so this improvement is only for a back to back deployments?

File: lib/templates/deploy.batch.go.template [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+:: 2. KuduSync
+ECHO Prepare source files for build
+IF /I "%IN_PLACE_DEPLOYMENT%" NEQ "1" (
+  call :ExecuteCmd "%KUDU_SYNC_CMD%" -v 50 -f "%DEPLOYMENT_SOURCE%" -t "%GOAZUREAPP%" -n "%NEXT_MANIFEST_PATH%" -p "%PREVIOUS_MANIFEST_PATH%" -i ".git;.hg;.deployment;deploy.cmd"
+  IF !ERRORLEVEL! NEQ 0 goto error
+)
+
+:: 3. Build
+ECHO Copy static files to deployment
+call :ExecuteCmd "%KUDU_SYNC_CMD%" -v 50 -f "%GOAZUREAPP%" -t "%DEPLOYMENT_TARGET%" -n "%NEXT_MANIFEST_PATH%-deploy" -p "%PREVIOUS_MANIFEST_PATH%-deploy" -i ".git;.hg;.deployment;deploy.cmd;*.go"
```
Comment (amitapl) [inline lib/templates/deploy.batch.go.template]: this kudusync invacation should use the usual manifest path the previous one should use a non-regular one.
(think of a scenario where you deployed go and then node app...)

File: lib/templates/deploy.batch.go.template [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    ECHO Creating %GOPATH%\src
+    MKDIR "%GOPATH%\src"
+)
+
+ECHO Creating %GOAZUREAPP%
+MKDIR %GOAZUREAPP%
+
+:: 2. KuduSync
+ECHO Prepare source files for build
+IF /I "%IN_PLACE_DEPLOYMENT%" NEQ "1" (
```
Comment (amitapl) [inline lib/templates/deploy.batch.go.template]: when %IN_PLACE_DEPLOYMENT% is set we should probably fail as it is not supported for go.

**Comment by shrimpy** [comment]
@amitapl 

RE:
as this is a temporary directory it will get cleaned up.
so this improvement is only for a back to back deployments?

kind of yes to your question.
for back to back deployment, it will definitively help for perf, since we only copied the changes. 
but depends on how often we clean up temp folder, if temp folder will be there for a a while, will be good for future deployment too.

**Comment by shrimpy** [comment]
Discuss with @amitapl, better approach would be use symlink for Go workspace structure. I am exploring the possibility.  

At the mean time i will continue what i am doing right now, have the workspace in temp folder. Will create a separate issue to improve it later.

File: lib/templates/deploy.batch.go.template [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+SET GOPATH=%DEPLOYMENT_TEMP%\gopath
+SET GOEXE="%GOROOT%\bin\go.exe"
+SET FOLDERNAME=azureapp
+SET GOAZUREAPP=%DEPLOYMENT_TEMP%\gopath\src\%FOLDERNAME%
+
+IF EXIST %GOPATH% (
+    ECHO GOPATH already exist %GOPATH%
+    
+    ECHO Removing %GOAZUREAPP%
+    RMDIR /S /Q %GOAZUREAPP%
```
Comment (amitapl) [inline lib/templates/deploy.batch.go.template]: I thought you said you dont need this?

**Comment by amitapl** [comment]
Awesome! :shipit:

File: lib/templates/deploy.batch.go.template (hunk 3 above)
Comment (shrimpy) [inline lib/templates/deploy.batch.go.template]: right.. will remove it. thanks!

**Pull request merged**